brandon_tan/partner_selector.py

```diff
@@ -0,0 +1,483 @@
+import pandas as pd
+import numpy as np
+
+import matplotlib.pyplot as plt
+
+from itertools import combinations
+
+from statsmodels.distributions.empirical_distribution import ECDF
+
+from math import comb
+
+from tqdm import tqdm
+
+
+class PartnerSelector:
+    """
+    Implementation of the Partner Selection Framework.
+    Hudson and Thames Skilling Challenge, March 2021.
+    Traditional Approach, extended approach and geometric approach are implemented.
+
+    https://www.econstor.eu/bitstream/10419/147450/1/870932616.pdf
+    """
+
+    def __init__(self, price_data: pd.DataFrame, target_ticker: str, no_of_partners: int = 3, top_n_corr: int = 50):
+        """
+        Constructor
+
+        Initialise the parter selector object.
+
+        :param price_data: Consumes a price dataframe provided by the user. Index by datetime and columns are tickers. Implicitly, the columns will define the universe that the partners will be found in.
+        :type price_data: pd.DataFrame
+        :param target_ticker: The target ticker that the user would like to create the grouping around.
+        :type price_data: str
+        :param no_of_partners: The number of partners to be identified.
+        :type price_data: int, optional
+        :param top_n_corr: The number of top n correlated securities to be used for partner selection process.
+        :type top_n_corr: int, optional
+        """
+
+        self.price_data = price_data
+        self.target_ticker = target_ticker
+        self.no_of_partners = no_of_partners
+        self.top_n_corr = top_n_corr
+        self.partner_data = None
+        self.best_Q = None
+        self._prepare_data()
+
+    def select(self, method: str, estimator: int = 1) -> list:
+        """
+        The main step of the framework. Based on the method that the user would like, we identify the
+        best partners based on the target security.
+
+        :param method: Either traditional, extended or geometric method as outlined in the paper.
+        :type method: str
+        :param estimator: The extended method has 3 estimators. This parameter identifies which estimators to be used. Defaults to 1.
+        :type estimator: int, optional
+        :raises Exception: In the event an invalid method is chosen, error will be raised.
+        :return: The best partners based on user chosen method will be returned in a list format.
+        :rtype: list
+        """
+
+        method = method.lower()
+        if method == 'traditional':
```

Comment:
I see... Okay got it! 

My initial thought was that it might be easier for users to remember the syntax ".select()" alone, and refer to documentation for what are the various types.